File: src/my_robot_tutorials/scripts/multiply_two_numbers_server.py
#!/usr/bin/env python3

import rospy
from my_robot_tutorials.srv import MultiplyTwoNumbers, MultiplyTwoNumbersRequest, MultiplyTwoNumbersResponse

SERVICE_NAME = "multiply_two_numbers"

def multiply(request: MultiplyTwoNumbersRequest) -> MultiplyTwoNumbersResponse:
    response = MultiplyTwoNumbersResponse()
    response.result = request.a * request.b
    return response


def main():
    rospy.init_node(f"{SERVICE_NAME}_server")
    rospy.Service(SERVICE_NAME, MultiplyTwoNumbers, multiply)
    rospy.loginfo("Server has been started.")
    rospy.loginfo("Ready to multiply two numbers.")
    rospy.spin()


if __name__ == "__main__":
    main()

File: src/my_robot_tutorials/src/multiply_two_numbers_server.cpp
#include <string>

#include <ros/ros.h>
#include "my_robot_tutorials/MultiplyTwoNumbers.h"

const std::string SERVICE_NAME = "multiply_two_numbers";

bool multiply(my_robot_tutorials::MultiplyTwoNumbers::Request &request,
              my_robot_tutorials::MultiplyTwoNumbers::Response &response) {
    response.result = request.a * request.b;
    return true;
}

int main(int argc, char **argv) {
    ros::init(argc, argv, (SERVICE_NAME + "_server").c_str());
    auto nh = ros::NodeHandle();

    // Save to a variable, even though we don't use it.
    // This is due to C++ RAII (Resource Acquisition Is Initialization) mechanism,
    // which automatically destructs the resource if it's not save to a variable.
    // By saving to a variable, the resource will not be destroyed until
    // the variable goes out of scope.
    // Similar mechanism happens with Rust.
    auto sub = nh.advertiseService(SERVICE_NAME.c_str(), multiply);

    ROS_INFO("Server has been started.");
    ROS_INFO("Ready to multiply two numbers.");

    ros::spin();
}
